python/ray/tests/kuberay/test_autoscaling_e2e.py

```diff
@@ -0,0 +1,243 @@
+        config["spec"]["workerGroupSpecs"][0]["maxReplicas"] = max_replicas
+
+        return config
+
+    def _apply_ray_cr(self, min_replicas=0, max_replicas=300, replicas=0) -> None:
+        """Apply Ray CR config yaml, with configurable replica fields for the single
+        workerGroup."""
+        with tempfile.NamedTemporaryFile("w") as config_file:
+            cr_config = self._get_ray_cr_config(
+                min_replicas=min_replicas, max_replicas=max_replicas, replicas=replicas
+            )
+            yaml.dump(cr_config, config_file)
+            config_file.flush()
+            subprocess.check_call(["kubectl", "apply", "-f", config_file.name])
+
+    def testAutoscaling(self):
+        """Test the following behaviors:
+
+        1. Spinning up a Ray cluster
+        2. Scaling up a Ray worker via autoscaler.sdk.request_resources()
+        3. Scaling up by updating the CRD's minReplicas
+        4. Scaling down by removing the resource request and reducing maxReplicas
+
+        Items 1. and 2. protect the example in the documentation.
+        Items 3. and 4. protect the autoscaler's ability to respond to Ray CR update.
+
+        Resources requested by this test are safely within the bounds of an m5.xlarge
+        instance.
+        - One Ray head pod
+            - Autoscaler: .5 CPU, .5 Gi memory
+            - Ray node: .5 CPU, .5 Gi memeory
+        - Two Worker pods
+            - Ray node: .5 CPU, .5 Gi memory
+        Total: 2 CPU, 2 Gi memory.
+
+        Including operator and system pods, the total CPU requested is just over 3.
+        """
+        # Cluster-creation
+        logger.info("Creating a RayCluster with no worker pods.")
+        self._apply_ray_cr(min_replicas=0, replicas=0)
+
+        logger.info("Confirming presence of head.")
+        wait_for_pods(goal_num_pods=1, namespace="default")
+        head_pod = get_pod(
+            pod_name_filter="raycluster-complete-head", namespace="default"
+        )
+
+        logger.info("Waiting for head pod to start Running.")
+        wait_for_pod_to_start(head_pod, namespace="default")
+        logger.info("Confirming Ray is up on the head pod.")
+        wait_for_ray_health(head_pod, namespace="default")
+
+        # Scale-up
+        logger.info("Scaling up to one worker via Ray resource request.")
+        scale_script = (
+            "import ray;"
+            'ray.init("auto");'
+            "ray.autoscaler.sdk.request_resources(num_cpus=2)"
+        )
+        kubectl_exec(
+            command=["python", "-c", scale_script], pod=head_pod, namespace="default"
+        )
+        logger.info("Confirming number of workers.")
+        wait_for_pods(goal_num_pods=2, namespace="default")
+
+        logger.info("Scaling up to two workers by editing minReplicas.")
+        # (replicas=1 reflects the current number of workers)
+        self._apply_ray_cr(min_replicas=2, replicas=1)
```

Comment:
minor and unrelated: I was thinking whether we should allow replicas  < min or > max. This is reflecting the desired state, then it doesn't make sense. For the real world cases, we can retrieve status from cluster.status field.  WDTY?

Reply:
There is definitely some complexity here, in particular with regards to maintaining consistent views of state in the operator and autoscaler.
I would definitely say that disallowing `replicas < minReplicas` in spec makes sense. I'm also not familiar with how the KubeRay operator uses minReplicas.

@sriram-anyscale has some ideas on these issues and related ones -- I think we should follow up with discussion on Slack and/or kuberay github.

Reply:
It's definitely a good idea to have the status field reflect at least the number of workers of each type.

It may be also useful for it to contain the entire list of worker pod names -- Ray workers are stateful and it's useful for autoscalers to have a consistent source of truth on which workers are in the cluster. Though again, we should probably have a dedicated discussion for these things. 

I've updated the `_apply_ray_cr` function to validate the `replicas` field on an existing CR when appropriate (to check that the autoscaler has set the replicas correctly on the CR).